### PR TITLE
Simplify Iterator::{min, max}

### PR DESCRIPTION
**Pull request by timvermeulen**
This PR simplifies the `select_fold1` helper method used to implmement `Iterator::{min, min_by, min_by_key, max, max_by, max_by_key}` by removing the projection argument, which was only used by the implementations of `min_by_key` and `max_by_key`.

I also added tests to ensure that the stability as mentioned in the comments of `min` and `max` is preserved, and fixed the `iter::{bench_max, bench_max_by_key}` benchmarks which the compiler presumably was able to collapse into closed-form expressions. None of the benchmark results were impacted, I suspect their generated assembly didn't change.

### CODE REVIEW

**Comment by cramertj** [comment]
Reassigning to someone on t-libs:

r? @sfackler

File: src/libcore/iter/traits/iterator.rs [hunk 1]
```diff
@@ -2038,12 +2034,8 @@ pub trait Iterator {
     #[stable(feature = "rust1", since = "1.0.0")]
     fn min(self) -> Option<Self::Item> where Self: Sized, Self::Item: Ord
     {
-        select_fold1(self,
-                     |_| (),
-                     // only switch to y if it is strictly smaller, to
-                     // preserve stability.
-                     |_, x, _, y| *x > *y)
-            .map(|(_, x)| x)
+        // only switch to y if it is strictly smaller, to preserve stability.
```
Comment (sfackler) [inline src/libcore/iter/traits/iterator.rs]: Can we just have this delegate to min_by?

Comment (timvermeulen) [reply src/libcore/iter/traits/iterator.rs]: We indeed can, I only considered forwarding to `min_by_key` 🙂 Thanks!

**Comment by sfackler** [comment]
@bors r+

Thanks!

**Pull request merged**